rbd: add persistent-cache command group

CongMinYin opened the pull request:
Update command for persistent cache.
`invalidate` command can discard cache and reset image cache state. If
cache file is lost, users can use this command to reset pwl image cache
state, which avoids the inability to use the cache due to file loss.

`flush `command can write cache files back to the osd. Explicitly
write back data.

update `help.t` and document.

Signed-off-by: Yin Congmin <congmin.yin@intel.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

CongMinYin commented:
@ideepika This is an urgent bug fix, if users uses RWL.

trociny reviewed (commented):
@CongMinYin  Could you please explain what is going on here and what expectations we should have about the cache auto recovery?

I mean that loosing persistent cache most likely means loosing some data. And it might be the case here? And we just silently reset the state without user noticing something was wrong? Shouldn't we be more careful here and ask the user to resolve this manually (may be additionally providing some tools and commands to mark the cache clean manually)?

I mean that it might be ok for the situation like emulated in the test just to reset the state (though I am not sure right now, because I don't have a good understanding what is going on). But in more general case, could the same issue be caused e.g. by the cache device not being mounted after the host reboot, or mounted with wrong permissions?

CongMinYin commented:
@trociny In fact, I find this bug is introduced by me https://github.com/ceph/ceph/pull/41968/commits/2cd9881116a47794051615fcf0a95a5e879ae7e6 . Maybe it is not a bug.  Instead, we need to provide users some command like `reset cache state`  as you said when they can't handle issue by themselves. The most primitive code logic directly ignores this error and initializes it. I also held the same view with you at the beginning: we can't handle it silently and let the user decide. I think "cache file lost" is a wrong state then add `return` in previous commit. Because there may be many kinds of situations, users can handle some of these:

- deveice not mount or directory configure wrong.
- cache file in another host and is waiting for write back.
- cache file really lose.
- ...

Your comments made me realize that not only the cache file lose error will make the image unusable. If there is a cache file, but the cache file data is corrupted, or version check or some value check is wrong after opening, the image can not be used until the user solves it. You provided a good idea. Maybe we should provide some commands. Users may be able to solve some problems through their own operations, such as mounting errors, cache files not written back on another host, etc. However, if the file is actually lost or corrupted, maybe we should provide some commands, such as `rbd pwl reset cache state`. I will do this.

trociny commented:
I don't like very much the command name: `rbd pwl reset cache state`.

Taking that the related config variables names are `rbd_persistent_cache_*`, I suggest the command name `rbd persistent-cache reset {image_spec}` or `rbd persistent-cache state reset {image_spec}` (though the last one looks too long).

Also I can imagine we might want in the future `rbd persistent-cache flush {image_spec}`, which would try to flush the cache. If the image is in use currently, it would just ask the lock owner to flush. If it is not used it would try to flush itself. With some special option (or as a separate command) it could even try to proceed when the cache is damaged trying to flush as much data as possible, probably providing information about damaged blocks (if it is possible).

And we could have some config variable (disabled by default) which would allow librbd to autorecover/reset the image cache on image opening even if the cache is damaged. It may be useful for the cases when for some reasons users prefer potential data lost instead of downtime resolving the issue manually.

CongMinYin commented:
@trociny When I almost finished the implementation of image cache `reset` and `flush` commands, I found that RBD commands can achieve the same effect~
```
rbd image-meta remove  test/image1 .librbd/image_cache_state
```
This command could flush cache to osd then reset image cache state. So, I think just updating error ouput and documentation is ok. The issue https://tracker.ceph.com/issues/53508 I create could be solved by this command. Sorry about that I don't know enough about RBD commands and PWL.

I will update later in this PR.

trociny commented:
@CongMinYin Although I am perfectly fine with just documenting how the cache may be reset with `rbd image-meta remove` command, I still think having a specific command(s) for manipulating the cache would be nice.

In future you might want to do some additional manipulations on reset, and all these details could be hidden from a user behind `rbd persistent-cache reset` command.

Note, we already have a similar situation with `rbd config image ...`  commands, which also manipulate with image meta, and `rbd image-meta ...` could be used instead.

I don't say you have to start working on this right now or even in near future. This may be just a TODO item.

CongMinYin commented:
@trociny OK, thanks your advice. Then I will continue to do this work now. Establish a command set of persistent cache to expand more commands in the future. Updated issue to explain the current solution.

CongMinYin commented:
@trociny I found there's alreay an persistent-cache command, it is named `image-cache invalidate`, which is implement by former author `Xiaoyan`.  It can validate cache file and then clean cache state. If the cache file lose, this command also can reset cache state . I changed it to `persistent-cache`  to avoid misunderstanding. I implemented 3 commands including `get` `invalidate` `flush`. Regarding `invalidate` and `flush` commands, I am worried that it will affect other caches in the future. Although there are only persistent-cache and rbd cache now.  Because they just call the `invalidate` and `flush` operations at the image level.  Perhaps the command can be explained in the document? The document will be updated later.

simple questions: Do I need to call `image.close() `explicitly? I found that many action files are not called.

trociny commented:
Ah, indeed, we already have `rbd image-cache invalidate`.

An open questions:  Do we really need to rename it to `persistent-cache`? Don't we want `image.invalidate_cache()` be generic to work for all types of caches (including the objecter cache)? I don't see reasons why it couldn't be. The same is for `flush` and status.

And if we still decide to rename it to `persistent-cache`, shouldn't we leave `image-cache invalidate` alisas? We usually did it marking as deprecated, so user scripts would still work if they used it. Not sure we need it in this case as the functionality is rather new.

And I don't like very much `get` command name. I would prefer `status`.

Also, note if you change CLI you also need to update `src/test/cli/rbd/help.t` (see  jenkins failure).

And I see `rbd image-cache invalidate` is already documented in doc/rbd/rbd-persistent-write-back-cache.rst, so it will need to be updated.

trociny commented:
@idryomov Do you have an opinion about (persistent) cache CLI?

trociny commented:
> simple questions: Do I need to call image.close() explicitly? I found that many action files are not called.

It is not necessary unless you want to control explicitly when exactly the image is closed (i.e. not when leaving the scope).

trociny commented:
And as a note. We already display the image status in `rbd status` command, though I don't mind having a separate `rbd ...-cache status` command.

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,234 @@
+    }
+    std::cout << image_cache_str << std::endl;
+  } else {
+    std::cout << "rbd: persistent cache is clean." << std::endl;
```
This is misleading because if `RBD_FEATURE_DIRTY_CACHE` is not set the cache may not even be enabled.

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,234 @@
+      std::cerr << "rbd: failed to get image cache state: "
+                << cpp_strerror(r) << std::endl;
+    }
+    std::cout << image_cache_str << std::endl;
```
Does this bring any value?  The exact same string is already output by `rbd status`.

Rather than adding a duplicate command I'd rather we try and address the main issue with image cache status -- the fact that the report is borderline useless and causes more harm than good because it is not updated while the image is open (https://tracker.ceph.com/issues/50614).

CongMinYin replied:
sorry, I only recently found the confusion of status information. I added `TODO` to this PR and I will resolve it in the near future. thanks for reminding.

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,234 @@
+  return r;
+}
+
+static int do_persistent_cache_discard(librbd::Image& image)
```
I would stick to the name of the command and the actual method: invalidate.

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,234 @@
+      return r;
+    }
+  } else {
+    std::cout << "rbd: persistent cache is already clean." << std::endl;
```
Same thing, the cache may just be disabled here.

idryomov commented:
> An open questions: Do we really need to rename it to `persistent-cache`? Don't we want `image.invalidate_cache()` be generic to work for all types of caches (including the objecter cache)? I don't see reasons why it couldn't be. The same is for `flush` and status.

The objecter cache is per-instance whereas the PWL cache is global (ignoring the corner case of the same image being used on different machines).  For the objecter cache, how would one express which instance is to be invalidated via this CLI?

IIRC it is already possible to flush and/or invalidate the objecter cache via the per-instance admin socket so I'm leaning towards renaming to `persistent-cache` to make it more obvious to users.

> 
> And if we still decide to rename it to `persistent-cache`, shouldn't we leave `image-cache invalidate` alisas? We usually did it marking as deprecated, so user scripts would still work if they used it. Not sure we need it in this case as the functionality is rather new.

Normally I would say yes, an alias is a must, but in this case we can probably do without it.  Persistent cache still has a lot of sharp edges and I doubt it has grown many real world users.  The only existing command in the group is `invalidate` which is (hopefully!) not something one would put in a script.

> 
> And I don't like very much `get` command name. I would prefer `status`.

Agreed.  We already have `rbd status` so I would drop it from here.

idryomov reviewed (commented):
It seems like there would be two separate things being done here:
- rename `image-cache` -> `persistent-cache`
- addition of `persistent-cache flush` command

Please split the change into two commits appropriately.

trociny commented:
> IIRC it is already possible to flush and/or invalidate the objecter cache via the per-instance admin socket so I'm leaning towards renaming to persistent-cache to make it more obvious to users.

My idea was that we would have the same (`image-cache`) command which would behave differently depending on what cache is used. Just to keep the list of commands shorter.  But agree, it might be confusing. So I am fine with renaming to `persistent-cache`.

> Agreed. We already have rbd status so I would drop it from here.

Do you mean we don't need `rbd persistent-cache status` at all because we already have the information in `rbd status`? To me, having both would not hurt (I just did not like the name ``rbd persistent-cache get`) but I have no strong opinion here.

idryomov commented:
> Do you mean we don't need `rbd persistent-cache status` at all because we already have the information in `rbd status`? To me, having both would not hurt (I just did not like the name ``rbd persistent-cache get`) but I have no strong opinion here.

Yes -- at least until all it outputs is a raw JSON string, `rbd status` is enough.  If/when we get some reasonably informative output (up-to-date cache state, dirty, flushing and clean amounts, flushing speed, etc), we can revisit.

CongMinYin commented:
@idryomov @trociny 
The three commands originally corresponded to three commands with the same effect.

- `rbd status` => `rbd persistent-cache get`. Drop it as suggested
- `rbd image-cache invalidate`  => `rbd persistent-cache invalidate`. Rename to `rbd persistent-cache`, and `remove image-cache` because There are very few real users now.
- `rbd image-meta remove  test/image1 .librbd/image_cache_state` => `rbd persistent-cache flush`.  Add independent commands.
- update `documentation` and `src/test/cli/rbd/help.t`

There is also an issue need to do: https://tracker.ceph.com/issues/50614.
I will proceed to make these changes. If there are other suggestions, please let me know.

trociny commented:
> `rbd image-meta remove test/image1 .librbd/image_cache_state`

Ah, one thing I just noticed. It is not related to this PR, I just want to discuss. I see that we are using `.librbd` prefix for this metadata key. I was not reviewing a PR when it was added, and I may be just missing the reason. Anyway, some time ago for librbd "internal" metadata keys we agreed to use ".rbd" prefix [1]. Was some particular reason why we decided not to use it for the image cache state? Is there a chance we can change it to ".rbd"?

[1] https://github.com/ceph/ceph/blob/242e8f35e8a19bad0b1125bd500221b1273e0612/src/librbd/image/GetMetadataRequest.cc#L24

idryomov commented:
> Is there a chance we can change it to ".rbd"?

Yes, unless there was a particular reason, I think we need to rename while the cache is still open for surgery and backward compatibility is not much of a concern.  IIRC it would result in this key being skipped by deep-copy and mirroring which seems desired since the backing file itself is not copied or mirrored.

CongMinYin commented:
> 
> Ah, one thing I just noticed. It is not related to this PR, I just want to discuss. I see that we are using `.librbd` prefix for this metadata key. I was not reviewing a PR when it was added, and I may be just missing the reason. Anyway, some time ago for librbd "internal" metadata keys we agreed to use ".rbd" prefix [1]. Was some particular reason why we decided not to use it for the image cache state? Is there a chance we can change it to ".rbd"?
> 
Thank you for your information, in order to be consistent, I think it needs to be changed to `.rbd`. Add to my TODO list.

CongMinYin commented:
@trociny Hi,  do you have any examples of other meta data key. I can't find other key has prefix “`.rbd`”.  It is more like a key without prefix spliced together with a prefix `.rbd`. But I'm not sure, because I can't track this sentence with gdb. Could you please find an example of another key for me to refer to? I mean, should the prefix be removed for this key, just like `IMAGE_CACHE_STATE = "image_cache_state"`?

https://github.com/ceph/ceph/blob/242e8f35e8a19bad0b1125bd500221b1273e0612/src/cls/rbd/cls_rbd.cc#L3711


In addition, there are some codes in PWL using `image cache`, which is easy to misunderstand. The full name of `PWL` is `persistent write log`, or we can call it `persistent cache`. It already has two names. Other names will make people more confusing. I want to change the naming of `image cache`, including the naming of keys, what is your suggestion.

trociny commented:
A special handling for ".rbd" prefixes was added when we were adding snapshot mirroring [1], at that time we decided that we might not want mirror some image metadata  and that would use ".rbd" prefix. At that time we didn't have yet any metadata at had to be handled that way. The first example is probably ".rbd_encryption" prefixed keys that are going to be added with encrypted clones support [2].

As for `persistent cache` vs `persistent write log` or `pwl` naming. I don't have a definite opinion. `persistent cache` looks good and preferable to me and it would be inline with `rbd_persistent_cache_*` configuration parameters we already have. If you have an example when it may be confusing please provide more details.

[1] https://github.com/ceph/ceph/pull/33490
[2] https://github.com/ceph/ceph/pull/40363/

CongMinYin commented:
remove commit about "[correct cache state](https://github.com/ceph/ceph/pull/44236/commits/7b1422c38e06c2f6a8ee9ba5323ec8741b28c78f)".
create a new pr to correct cache state: https://github.com/ceph/ceph/pull/45660

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+
+Shell::Action action_invalidate(
+  {"persistent-cache", "invalidate"}, {},
+  "Invalidate existing / dirty image cache. Reset persistent cache state.", "",
```
```suggestion
  "Invalidate existing / dirty persistent cache.", "",
```

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+
+} // anonymous namespace
+
+static int do_persistent_cache_invalidate(librbd::Image& image)
+{
```
```suggestion
static int do_persistent_cache_invalidate(librbd::Image& image) {
```

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+  return r;
+}
+
+static int do_persistent_cache_flush(librbd::Image& image)
+{
```
```suggestion
static int do_persistent_cache_flush(librbd::Image& image) {
```

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+    r = image.invalidate_cache();
+    if(r < 0) {
+      std::cerr << "failed to invalidate persistent cache: "
+                << cpp_strerror(r) << std::endl;
```
Remove this error message, there already is one in `execute_invalidate`.

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+  if (features & RBD_FEATURE_DIRTY_CACHE) {
+    std::cout << "flushing... " << std::endl;
+    r = image.flush();
+    if(r < 0) {
```
```suggestion
    if (r < 0) {
```

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+    return r;
+  }
+
+  return r;
```
```suggestion
  return 0;
```

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+    return r;
+  }
+
+  if (features & RBD_FEATURE_DIRTY_CACHE) {
```
Why did you add the feature bit check here?

It looks like it matters in case the cache status metadata key is present but the feature bit is not set.  In that case with the old code we would have removed the key but with the new code it is left around.  Or am I misreading this?

CongMinYin replied:
You're right. I should not make this change. In any case, the invalidate command should work.

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+
+  r = do_persistent_cache_flush(image);
+  if (r < 0) {
+    std::cerr << "rbd: flushing persistent cache state failed: "
```
```suggestion
    std::cerr << "rbd: flushing persistent cache failed: "
```

idryomov commented on src/tools/rbd/action/PersistentCache.cc:
```diff
@@ -0,0 +1,159 @@
+    r = image.flush();
+    if(r < 0) {
+      std::cerr << "failed to flush persistent cache: "
+                << cpp_strerror(r) << std::endl;
```
Remove this error message, there already is one in `execute_flush`.

idryomov reviewed (changes requested):
Don't forget to update documentation and `src/test/cli/rbd/help.t` as mentioned in https://github.com/ceph/ceph/pull/44236#issuecomment-997555763.

CongMinYin commented:
jenkins retest this please

CongMinYin commented:
jenkins test make check

CongMinYin commented:
make check failed due to `help.t` ctest failed, Does anyone know why? I have no experience in `help.t`

idryomov commented:
I incorporated the (slightly modified) documentation and `help.t` snippets into respective commits and fixed a couple of related issues found in play testing.  @CongMinYin please take a look.

CongMinYin commented:
> I incorporated the (slightly modified) documentation and `help.t` snippets into respective commits and fixed a couple of related issues found in play testing. @CongMinYin please take a look.

LGTM. Thanks.

idryomov commented:
Rebased to resolve conflicts with https://github.com/ceph/ceph/pull/45684.

idryomov commented:
No related failures:

https://pulpito.ceph.com/dis-2022-04-12_23:22:20-rbd-wip-dis-testing-1-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-04-13_07:15:09-rbd-wip-dis-testing-1-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-04-13_09:06:29-rbd-wip-dis-testing-1-distro-default-smithi/
